### PR TITLE
New examples: filter positive elements of a vector.

### PR DESCRIPTION
**Pull request by claudemarche**
The example inspired from P. Muller

### CODE REVIEW

**Comment by claudemarche** [comment]
Uses a logic function and a two lemmas

File: creusot/tests/should_succeed/filter_positive.rs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+#[logic]
+#[requires(j <= k)]
+#[ensures(num_of_pos(i,j,t) <= num_of_pos(i,k,t))]
+#[variant(k-j)]
+fn lemma_num_of_pos_increasing(i: Int, j: Int, k: Int, t: Seq<i32>) {
+    pearlite! {
+        if j < k {
+            lemma_num_of_pos_increasing(i,j+1,k,t)
+        }
+        else {  // Creusot panics if no else given...
```
Comment (jhjourdan) [inline creusot/tests/should_succeed/filter_positive.rs:67]: @xldenis, perhaps you could fix that?

File: creusot/tests/should_succeed/filter_positive.rs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+// lemma: `num_of_pos` is strictly increasing when a positive element
+// is met
+#[logic]
+#[requires(0 <= i && i < t.len())]
+#[requires(@t[i] > 0)]
+#[ensures(num_of_pos(0,i,t) < num_of_pos(0,i+1,t))]
+fn lemma_num_of_pos_strictly_increasing(i: Int, t: Seq<i32>) {}
+
+// the main function to proof
```
Comment (jhjourdan) [inline creusot/tests/should_succeed/filter_positive.rs:81]:
```suggestion
// the main function to prove
```

File: creusot/tests/should_succeed/filter_positive.rs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            count += 1;
+        }
+        i += 1;
+    }
+    return u;
+}
+
+// main program, for testing purpose
+// commented out here, because Creusot complains
+fn main() {
```
Comment (jhjourdan) [inline creusot/tests/should_succeed/filter_positive.rs:122]: Have you tried with a `#[trusted]` attribute, and a `#[require(false)]` precondition, to disable proof checking and ensure that this function is not called?

File: creusot/tests/should_succeed/filter_positive.rs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    #[invariant(loop_bounds, 0 <= @i && @i <= (@t).len())]
+    #[invariant(count_bounds, 0 <= @count && @count <= @i)]
+    #[invariant(num, @count === num_of_pos(0,@i,@t))]
+    #[invariant(ulength, (@u).len() === num_of_pos(0,(@t).len(),@t))]
+    //#[variant((@t).len() - @i)]
+    while i < t.len() {
+        if t[i] > 0 {
+            // the tricky assertions, that needs lemmas
+            proof_assert! { {lemma_num_of_pos_strictly_increasing(@i,@u);
+            num_of_pos(0,@i,@t) < num_of_pos(0,@i+1,@t) } };
```
Comment (jhjourdan) [inline creusot/tests/should_succeed/filter_positive.rs:109]:
```suggestion
                num_of_pos(0,@i,@t) < num_of_pos(0,@i+1,@t) } };
```

File: creusot/tests/should_succeed/filter_positive.rs [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+            proof_assert! { {lemma_num_of_pos_increasing(0,@i+1,(@t).len(),@t);
```
Comment (jhjourdan) [inline creusot/tests/should_succeed/filter_positive.rs:110]:
```suggestion
                proof_assert! { {lemma_num_of_pos_increasing(0,@i+1,(@t).len(),@t);
```

File: creusot/tests/should_succeed/filter_positive.rs (hunk 1 above)
Comment (xldenis) [reply creusot/tests/should_succeed/filter_positive.rs:67]: Is that valid why3 syntax?

File: creusot/tests/should_succeed/filter_positive.rs [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+#[requires(@t[i] > 0)]
+#[ensures(num_of_pos(0,i,t) < num_of_pos(0,i+1,t))]
+fn lemma_num_of_pos_strictly_increasing(i: Int, t: Seq<i32>) {}
+
+// the main function to proof
+fn m(t: Vec<i32>) -> Vec<i32> {
+    let mut count: usize = 0;
+    let mut i: usize = 0;
+    #[invariant(loop_bounds, 0 <= @i && @i <= (@t).len())]
+    #[invariant(count_bounds, 0 <= @count && @count <= @i)]
```
Comment (jhjourdan) [inline creusot/tests/should_succeed/filter_positive.rs:86]: Why is this invariant needed?

File: creusot/tests/should_succeed/filter_positive.rs [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        }
+        i += 1;
+    }
+    let mut u: Vec<i32> = vec::from_elem(0, count);
+    //let mut u : Vec<i32> = vec![0;count];
+    count = 0;
+
+    i = 0;
+    #[invariant(loop_bounds, 0 <= @i && @i <= (@t).len())]
+    #[invariant(count_bounds, 0 <= @count && @count <= @i)]
```
Comment (jhjourdan) [inline creusot/tests/should_succeed/filter_positive.rs:101]: Same question here.

**Review by jhjourdan — APPROVED** [review]

File: creusot/tests/should_succeed/filter_positive.rs (hunk 3 above)
Comment (xldenis) [reply creusot/tests/should_succeed/filter_positive.rs:122]: That should work here, there is still an issue in general where closures can escape `#[trusted]` but the commented code doesn't make use of them.

Comment (claudemarche) [reply creusot/tests/should_succeed/filter_positive.rs:122]: No i didn't. Anyway, I'd really would like to run Creusot and compile with rustc on the same file, without needing to comment things for Creusot.

File: creusot/tests/should_succeed/filter_positive.rs (hunk 4 above)
Comment (claudemarche) [reply creusot/tests/should_succeed/filter_positive.rs:109]: not my fault, it is rustfmt

File: creusot/tests/should_succeed/filter_positive.rs (hunk 5 above)
Comment (claudemarche) [reply creusot/tests/should_succeed/filter_positive.rs:110]: rustfmt...

File: creusot/tests/should_succeed/filter_positive.rs (hunk 6 above)
Comment (claudemarche) [reply creusot/tests/should_succeed/filter_positive.rs:86]: Well, count >= 0 might be enough, indeed

File: creusot/tests/should_succeed/filter_positive.rs (hunk 4 above)
Comment (xldenis) [reply creusot/tests/should_succeed/filter_positive.rs:109]: `rustfmt` doesn't touch anything inside of proc-macros like `proof_assert!` so you should / need to adjust it yourself.

File: creusot/tests/should_succeed/filter_positive.rs (hunk 7 above)
Comment (claudemarche) [reply creusot/tests/should_succeed/filter_positive.rs:101]: same answer here ;-)

Comment (jhjourdan) [reply creusot/tests/should_succeed/filter_positive.rs:101]: Ahah ! Same answer, then!

File: creusot/tests/should_succeed/filter_positive.rs (hunk 6 above)
Comment (jhjourdan) [reply creusot/tests/should_succeed/filter_positive.rs:86]: Actually, `count >= 0` is not needed, but `count <= i` is needed to check for overflows.

File: creusot/tests/should_succeed/filter_positive.rs (hunk 4 above)
Comment (jhjourdan) [reply creusot/tests/should_succeed/filter_positive.rs:109]: Apparently, here it does...

**Pull request merged**